bpo-39481: Implementation for PEP 585

gvanrossum opened the pull request:
I consider this a finished prototype now. We've got everything working, and then some (e.g. pickle), and while we found a fair number of minor issues with PEP 585, the main conclusion is that the PEP is quite reasonable to implement.

A number of suggestions for changes PEP 585 that have basically been vetted by @ambv can be found in https://github.com/python/peps/pull/1289.

~There are a few places where I still disagree with @ambv. In particular, I don't think it makes sense to even allow `isinstance(x, list[int])` and `issubclass(C, list[int])` -- IMO these should just raise TypeError, and that's what this PR implements. (If @ambv convinces me to change my mind, it's easy to revert that.)~ [**UPDATE:** @ambv agrees and has [updated the PEP](https://github.com/python/peps/pull/1311).]

There are also a few things that ought to be changed:

- [x] We should add a nice docstring to `GenericAlias`.
- [x] We made `io.IOBase` generic. This was a misunderstanding, and we'll revert that.
- [x] The `__parameters__` attribute can and probably should be computed lazily the first time it's needed.
- [ ] Maybe we still need a cache, like `typing.py` uses?
- [x] Where should the code live? For now I put it in `descrobject.c`, but that's debatable (it was just convenient).

<!-- issue-number: [bpo-39481](https://bugs.python.org/issue39481) -->
https://bugs.python.org/issue39481
<!-- /issue-number -->

gvanrossum committed:
PEP 585 step 2: Write a Py_GenericAlias() function that returns origin

gvanrossum committed:
PEP 585 step 3: Add Py_GenericAlias() as __class_getitem__ to tuple, list, dict, set, frozenset

gvanrossum committed:
PEP 585 step 4: Write some tests

gvanrossum committed:
Define a GenericAlias (proxy) object (with lots of TODOs still)

gvanrossum committed:
More tests

gvanrossum committed:
PEP 585 steps 5-6: Finish basic GenericAlias object and test

gvanrossum committed:
Fix bug with class methods

gvanrossum committed:
Define tp_new instead of tp_init

gvanrossum committed:
Test that list[int][int] fails

gvanrossum committed:
Test that class L(list): pass; L[int] works

emmatyping committed:
Implement PEP 585 for collections.deque and ... (#2)

... collections.defaultdict

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,152 @@ PyTypeObject PyProperty_Type = {
+    return PyObject_Call(alias->origin, args, kwds);
+}
+
+static const char* attr_exceptions[] = {
```
```suggestion
static const char* const attr_exceptions[] = {
```

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,152 @@ PyTypeObject PyProperty_Type = {
+{
+    gaobject *alias = (gaobject *)self;
+    if (alias->origin != NULL && PyUnicode_Check(name)) {
+        for (const char **p = attr_exceptions; ; p++) {
```
```suggestion
        for (const char * condt *p = attr_exceptions; ; p++) {
```

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,152 @@ PyTypeObject PyProperty_Type = {
+ga_getattro(PyObject *self, PyObject *name)
+{
+    gaobject *alias = (gaobject *)self;
+    if (alias->origin != NULL && PyUnicode_Check(name)) {
```
In what case `alias->origin == NULL`?

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,152 @@ PyTypeObject PyProperty_Type = {
+ga_mro_entries(PyObject *self, PyObject *args)
+{
+    gaobject *alias = (gaobject *)self;
+    return Py_BuildValue("(O)", alias->origin);
```
```suggestion
    return PyTuple_Pack(1, alias->origin);
```

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,152 @@ PyTypeObject PyProperty_Type = {
+};
+
+static PyMemberDef ga_members[] = {
+    {"__origin__", T_OBJECT, offsetof(gaobject, origin), READONLY},
```
Maybe `T_OBJECT_EX`?

gvanrossum replied:
I think not. The two attributes should never be allowed to be NULL. (This is a change from an earlier design.)

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,152 @@ PyTypeObject PyProperty_Type = {
+static PyObject *
+ga_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    if (kwds != NULL) {
```
```suggestion
    if (kwds != NULL && PyDict_GET_SIZE(kwds) != 0) {
```

serhiy-storchaka replied:
It is worth to use Argument Clinic which generates simpler code using non-public C API similar to the following:
```c
if (!_PyArg_NoKeywords("GenericAlias", kwds) ||
    !_PyArg_CheckPositional("GenericAlias", PyTuple_GET_SIZE(args), 2, 2))
{
    return NULL;
}
```

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,152 @@ PyTypeObject PyProperty_Type = {
+    if (kwds != NULL) {
+        return PyErr_Format(PyExc_TypeError, "GenericAlias does not support keyword arguments");
+    }
+    if (!PyTuple_Check(args) || PyTuple_Size(args) != 2) {
```
`PyTuple_Check(args)` is redundant. `args` is always a tuple.

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,152 @@ PyTypeObject PyProperty_Type = {
+{
+    PyObject *wrapper = NULL;
+    if (!PyTuple_Check(parameters)) {
+        wrapper = PyTuple_New(1);
```
```c
if (!PyTuple_Check(parameters)) {
    parameters = PyTuple_Pack(1, parameters);
    if (parameters == NULL) {
        return NULL;
    }
}
else {
    Py_INCREF(parameters);
}
```
Also add `Py_DECREF(parameters)` after `if (alias == NULL)` and remove `Py_INCREF(parameters)` below.

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,152 @@ PyTypeObject PyProperty_Type = {
+};
+
+static PyObject *
+ga_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
```
Do we need a `__new__` method? Why not set `tp_new = NULL`?

gvanrossum replied:
I first had a tp_init. Then calling GenericAlias() somehow worked and produced an object with NULL attributes (or maybe it was some more clever incantation, I forget, but I definitely saw it). I do want the class instantiable from Python (so it can potentially be used from typing.py). Using tp_new instead solved that problem.

gvanrossum committed:
Respond to Serhiy's code review

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

emmatyping committed:
PEP 585 steps 7-8: Implement repr for GenericAlias (#1)

By Ethan Smith.

emmatyping committed:
Add static keyword to ga_repr_item (#3)

gvanrossum committed:
Add GenericAlias to types.py

gvanrossum committed:
Add test for types.py

Also strengthen test for repr(MyList[int])

gvanrossum committed:
Fix tests that were sensitive to the presence of __class_getitem__

Note that UserDict and UserList are now generic

gvanrossum committed:
Fix crash in repr(tuple[...]) -- call PyErr_Clear()

emmatyping committed:
Make re.Match, re.Pattern, and io.IOBase generic (#4)

gvanrossum committed:
Restore parity between _io._IOBase and _pyio.IOBase

serhiy-storchaka commented on Lib/test/test_genericalias.py:
```diff
@@ -2,12 +2,14 @@
 
     def test_subscriptable(self):
-        for t in tuple, list, dict, set, frozenset, defaultdict, deque:
+        for t in tuple, list, dict, set, frozenset, defaultdict, deque, IOBase, Pattern, Match:
```
TODO: I suppose IOBase subclasses BufferedIOBase, RawIOBase and TextIOBase should not support indexing.

serhiy-storchaka replied:
There is also Python implementation `_pyio.IOBase`. I am not sure what to do with it. Maybe just ignore for now.

gvanrossum replied:
I added `__class_getitem__` to `_pyio.IOBase` in 35cd5f477c9da985880130d72297335e010450e4.

I'm not sure whether it matters much that we take the `__class_getitem__` away from the various subclasses (let the static type checkers worry about that).

gvanrossum committed:
Make contextlib.Abstract*ContextManager generic, with tests

gvanrossum committed:
Rename __parameters__ to __args__

gvanrossum committed:
Rename parameters to args

Keep not-yet-functional parameters, and add __parameters__ to match needed.

gvanrossum committed:
Properly implement __parameters__ as the tuple of args that are type vars

gvanrossum committed:
Set __orig_class__ if possible when instantiating

gvanrossum committed:
WIP: Tweak test_typing.py to pass with the new definition of __orig_class__

gvanrossum committed:
WIP: Implement ga_getitem so list[T][int] -> list[int]

However, dict[T, int][str] -> dict[str], which is wrong!
Similar, dict[T, T][str] fails but should return dict[str, str].
Finally, dict[T, T].__parameters__ -> (T, T) but should be (T,).

gvanrossum committed:
Exclude duplicates from __parameters__

Also move declaration of Py_GenericAliasType to descrobject.h.

gvanrossum committed:
Implement ga_getitem properly, with tests

gvanrossum committed:
Implement == and != for GA

emmatyping committed:
Make collections.abc.* generic (#5)

emmatyping committed:
Make isinstance and issubclass error for GenericAlias (#6)

gvanrossum committed:
Fix two more AssertIs calls in test_typing.py

gvanrossum committed:
Admit defeat -- comment out two expected TypeErrors in test_extended_generic_rules_subclassing

gvanrossum committed:
Fix whitespace to satisfy patchcheck.py

gvanrossum committed:
Fix test_site.py

gvanrossum committed:
Better way to fix test_site.py: don't import types from collections

emmatyping committed:
Use Py_ssize_t instead of int for indexing (#8)

emmatyping committed:
GenericAlias pickle support (#9)

emmatyping committed:
Make type generic (#7)

This change specifically allows type[int], but disallows str[int].

gvanrossum committed:
Remove a completed TODO comment

emmatyping committed:
Remove extraneous whitespace. (#10)

emmatyping committed:
Do not make IOBase generic (#11)

gvanrossum committed:
Remove __class_getitem__ from _pyio.py too; and don't test for IOBase

gvanrossum committed:
Merge branch 'master' into pep585

ambv commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,423 @@ PyTypeObject PyProperty_Type = {
+ga_instancecheck(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    return PyErr_Format(PyExc_TypeError,
+                        "TypeError: Subscripted generics cannot be used with class and instance checks",
```
I don't think this should spell out "TypeError: " explicitly, it produces this:
```pycon
>>> isinstance([], list[str])
...
TypeError: TypeError: Subscripted generics cannot be used with class and instance checks
```

I would also suggest this for the message which is consistent with the message for other invalid use:
"isinstance() arg 2 cannot be a parametrized generic"

ambv replied:
By the way, shouldn't we also disallow:
```pycon
>>> issubclass(list[str], list)
False
```

emmatyping replied:
Ah that's on me, I'll submit fixes for these.

emmatyping replied:
Okay hm, for the case of:

```
>>> issubclass(list[str], list)
```

I'm stumped. Perhaps this needs to be implemented in the builtin `isinstance`/`issubclass`? Otherwise I don't see how to prevent that pattern.

ambv commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,423 @@ PyTypeObject PyProperty_Type = {
+ga_subclasscheck(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    return PyErr_Format(PyExc_TypeError,
+                        "TypeError: Subscripted generics cannot be used with class and instance checks",
```
Ditto about the "TypeError: ".
Ditto the message would be:

"issubclass() arg 2 cannot be a parametrized generic"

ambv commented on Objects/descrobject.c:
```diff
@@ -1790,3 +1790,423 @@ PyTypeObject PyProperty_Type = {
+// - cache?
+PyTypeObject Py_GenericAliasType = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "types.GenericAlias",
```
I know it's late for this bikeshedding but I was always a bit puzzled by the name "GenericAlias". What is it aliasing?

Especially now that it lives in `types`, wouldn't "ParametrizedType" work better?

emmatyping replied:
I think "ParametrizedType" is both long and easy to misspell (I can already tell I will write "ParameterizedType" and be confused).

Things in `type` generally are `NameType`, like `FunctionType` or `CodeType`, so perhaps `GenericType`? I agree the word "alias" is confusing.

gvanrossum replied:
Agreed that GenericAlias is a lame name for the concept. It's easy to change now, but will be impossible to change once the PEP's been accepted, so it's appropriate to bikeshed about it (here, and maybe in the list). At first blush I do like GenericType.

gvanrossum replied:
I brought this up on python-dev and it looks like GenericAlias is here to stay.

emmatyping committed:
Make types.MappingProxyType generic (#13)

emmatyping committed:
Reword isinstance/subclass checks (#12)

gvanrossum committed:
Fix test_types (mappingproxy now has a __class_getitem__ method)

gvanrossum committed:
Merge remote-tracking branch 'upstream/master' into pep585

emmatyping committed:
Convert some already generic things to GenericAlias to be consistent (#14)

emmatyping committed:
Make __parameters__ lazy (#15)

Now we can also remove `__setstate__`.

gvanrossum committed:
Fix failing test_doctest -- we now find 821 tests

gvanrossum committed:
Don't import types in os.py -- it breaks test_site.py

gvanrossum committed:
Move GenericAlias definition to its own file

gvanrossum committed:
Add a hash function to GenericAlias

gvanrossum committed:
Make Union[list[T], int] work

This involves a rare change in typing.py.

emmatyping committed:
Add GenericAlias to Windows build (#16)

gvanrossum committed:
Merge remote-tracking branch 'upstream/master' into pep585

gvanrossum committed:
Run CI on PRs for pep585 branch

gvanrossum committed:
Merge branch 'master' into pep585

emmatyping committed:
Add docstring for GenericAlias (#17)

Co-authored-by: Guido van Rossum <guido@python.org>

rhettinger commented on Lib/collections/__init__.py:
```diff
@@ -38,6 +38,8 @@
 except ImportError:
     pass
 
+_GenericAlias = type(list[int])
```
Is this supposed to be here?  AFAICT it isn't referenced anywhere and the tests pass without it.

gvanrossum replied:
Good catch. I'm guessing this was used in an earlier version. I'll take it out.

gvanrossum committed:
Remove unneeded _GenericAlias definition from collections/__init__.py

rhettinger commented on Objects/tupleobject.c:
```diff
@@ -832,6 +832,7 @@ static PyMethodDef tuple_methods[] = {
     TUPLE___GETNEWARGS___METHODDEF
     TUPLE_INDEX_METHODDEF
     TUPLE_COUNT_METHODDEF
```
Since this method is going to become so pervasive, consider adding a tp_slot for it.

gvanrossum replied:
Hm, I don't know if it's worth it. There are currently just 7 occurrences in Object/ and another 7 in Modules/.

rhettinger replied:
It doesn't have to be decided now, you can switch to slots in the future if the balance tips the other way.

ilevkivskyi commented on Lib/os.py:
```diff
@@ -28,6 +28,8 @@
 
 from _collections_abc import _check_methods
 
+GenericAlias = type(list[int])
```
Why this module, and e.g. `_collections_abc` use this code but other use `import types; types.GenericAlias`?

gvanrossum replied:
The "right" way is to import it from `types`, but we can't use that here. The `os` module is part of a small set of modules that are imported at startup. The `types` module is not part of that set, and there's a unit test that checks this. (In test/test_site.py, in class `StartupImportTests`, look at ``test_startup_imports`.)

ilevkivskyi commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,447 @@
+is_typevar(PyObject *obj)
+{
+    PyTypeObject *type = Py_TYPE(obj);
+    return strcmp(type->tp_name, "TypeVar") == 0;
```
Maybe it would be better to also check that `TypeVar.__module__ == 'typing'`?

gvanrossum replied:
I was hoping I wouldn't have to write that code, but you're right. I'll add it.

gvanrossum replied:
Added.

ilevkivskyi reviewed (approved):
Thanks! I read through the code and didn't find anything that looks problematic.

serhiy-storchaka commented on Lib/_pyio.py:
```diff
@@ -8,6 +8,7 @@
 import errno
 import stat
 import sys
+import types
```
Not used.

serhiy-storchaka commented on Lib/_collections_abc.py:
```diff
@@ -110,6 +112,10 @@ def __subclasshook__(cls, C):
             return _check_methods(C, "__await__")
         return NotImplemented
 
+    def __class_getitem__(cls, item):
```
I am wondering whether we can just write
```python
__class_getitem__ = classmethod(GenericAlias)
```

gvanrossum replied:
Good catch! I tried without `classmethod()` and that didn't work. But this does. I'll update all occurrences.

gvanrossum committed:
Make is_typevar() harder to fool

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,447 @@
+    Py_XDECREF(module);
+    Py_XDECREF(r);
+    return err;
+}
```
It is not good to silence an arbitrary exception. It may be KeyboardInterrupt, MemoryError, RecursionError, etc, and in rare cases when it happens, you will get a non-reproducible change in behavior.

There are also possible crashes if one attribute raises an error and other executes Python code.

There are also possible crashes if `__module__` or `__qualname__` are not strings.

```suggestion
static int
ga_repr_item(_PyUnicodeWriter *writer, PyObject *p)
{
    _Py_IDENTIFIER(__module__);
    _Py_IDENTIFIER(__qualname__);
    _Py_IDENTIFIER(__origin__);
    _Py_IDENTIFIER(__args__);
    PyObject *qualname = NULL;
    PyObject *module = NULL;
    PyObject *r = NULL;
    PyObject *tmp;
    int err;

    if (p == Py_Ellipsis) {
        // The Ellipsis object
        r = PyUnicode_FromString("...");
        goto done;
    }

    if (_PyObject_LookupAttrId(p, &PyId___origin__, &tmp) < 0) {
        goto done;
    }
    if (tmp != NULL) {
        Py_DECREF(tmp);
        if (_PyObject_LookupAttrId(p, &PyId___args__, &tmp) < 0) {
            goto done;
        }
        if (tmp != NULL) {
            Py_DECREF(tmp);
            // It looks like a GenericAlias
            goto use_repr;
        }
    }

    if (_PyObject_LookupAttrId(p, &PyId___qualname__, &qualname) < 0) {
        goto done;
    }
    if (qualname == NULL) {
        goto use_repr;
    }
    if (_PyObject_LookupAttrId(p, &PyId___module__, &module) < 0) {
        goto done;
    }
    if (module == NULL || module == Py_None) {
        goto use_repr;
    }

    // Looks like a class
    if (PyUnicode_Check(module) &&
        _PyUnicode_EqualToASCIIString(module, "builtins"))
    {
        // builtins don't need a module name
        Py_INCREF(qualname);
        r = qualname;
        goto done;
    }
    else {
        r = PyUnicode_FromFormat("%S.%S", module, qualname);
        goto done;
    }

use_repr:
    r = PyObject_Repr(p);

done:
    Py_XDECREF(qualname);
    Py_XDECREF(module);
    if (r == NULL) {
        // error if any of the above PyObject_Repr/PyUnicode_From* fail
        err = -1;
    }
    else {
        err = _PyUnicodeWriter_WriteStr(writer, r);
        Py_DECREF(r);
    }
    return err;
}
```

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+        return 0;
+    }
+    int res = PyUnicode_Check(module)
+        && PyUnicode_CompareWithASCIIString(module, "typing") == 0;
```
```suggestion
        && _PyUnicode_EqualToASCIIString(module, "typing");
```
It is just a tiny bit more efficient.

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+    }
+    PyObject *module = PyObject_GetAttrString((PyObject *)type, "__module__");
+    if (module == NULL) {
+        // Oops.  (Bubbling up the error would just complicate the callers.)
```
Of, it is harder to fix. I'll do this later.

gvanrossum replied:
Thanks.

gvanrossum replied:
Never mind, I did it. Wasn't too complicated, but please check.

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+ga_repr(PyObject *self)
+{
+    gaobject *alias = (gaobject *)self;
+    Py_ssize_t len = PyTuple_Size(alias->args);
```
It is guaranteed that `alias->args` is a tuple and `PyTuple_Size()` never fails. I would use the macro `PyTuple_GET_SIZE` to underscore this fact. In any case you use `PyTuple_GET_ITEM` below.

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+        if (params == NULL) {
+            return NULL;
+        }
+        Py_DECREF(params);
```
Maybe call `make_parameters()` directly?
```suggestion
        alias->parameters = make_parameters(alias->args);
        if (alias->parameters == NULL) {
            return NULL;
        }
```

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+    gaobject *alias = (gaobject *)self;
+    // TODO: Hash in the hash for the origin
+    Py_hash_t h0 = PyObject_Hash(alias->origin);
+    Py_hash_t h1 = PyObject_Hash(alias->args);
```
Error handling. Hash function can execute Python code and it will crash if an error is set.
```suggestion
    Py_hash_t h0 = PyObject_Hash(alias->origin);
    if (h0 == -1) {
        return -1;
    }
    Py_hash_t h1 = PyObject_Hash(alias->args);
    if (h1 == -1) {
        return -1;
    }
```

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+    PyObject *obj = PyObject_Call(alias->origin, args, kwds);
+    if (obj != NULL) {
+        PyObject_SetAttrString(obj, "__orig_class__", self);
+        PyErr_Clear();
```
```suggestion
        if (PyObject_SetAttrString(obj, "__orig_class__", self) < 0) {
            if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
                Py_DECREF(obj);
                return NULL;
            }
            PyErr_Clear();
        }
```

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+            if (*p == NULL) {
+                return PyObject_GetAttr(alias->origin, name);
+            }
+            if (PyUnicode_CompareWithASCIIString(name, *p) == 0) {
```
Just a microoptimization.
```suggestion
            if (_PyUnicode_EqualToASCIIString(name, *p)) {
```

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+    if (eq == Py_False) {
+        return eq;
+    }
+    Py_DECREF(eq);
```
What if `PyObject_RichCompare()` returns false value but not `False`? I would use `PyObject_RichCompareBool()`. It would also make the code simpler.
```suggestion
    int eq = PyObject_RichCompareBool(aa->origin, bb->origin, Py_EQ);
    if (eq < 0) {
        return NULL;
    }
    if (!eq) {
        Py_RETURN_FALSE;
    }
```

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+{
+    return PyErr_Format(PyExc_TypeError,
+                        "isinstance() argument 2 cannot be a parameterized generic",
+                        self);
```
The format string does not use `self`. Maybe use `PyErr_SetString()`?

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+{
+    return PyErr_Format(PyExc_TypeError,
+                        "issubclass() argument 2 cannot be a parameterized generic",
+                        self);
```
Same as above.

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+    if (alias->parameters == NULL) {
+        alias->parameters = make_parameters(alias->args);
+        if (alias->parameters == NULL) {
+            Py_DECREF(alias->parameters);
```
Decrefing NULL.
```suggestion
        if (alias->parameters == NULL) {
```

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+ga_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    if (kwds != NULL && PyDict_GET_SIZE(kwds) != 0) {
+        return PyErr_Format(PyExc_TypeError, "GenericAlias does not support keyword arguments");
```
```suggestion
        return PyErr_SetString(PyExc_TypeError, "GenericAlias does not support keyword arguments");
```

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+        return PyErr_Format(PyExc_TypeError, "GenericAlias does not support keyword arguments");
+    }
+    if (PyTuple_Size(args) != 2) {
+        return PyErr_Format(PyExc_TypeError, "GenericAlias expects 2 positional arguments");
```
```suggestion
        return PyErr_SetString(PyExc_TypeError, "GenericAlias expects 2 positional arguments");
```

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -0,0 +1,455 @@
+    if (kwds != NULL && PyDict_GET_SIZE(kwds) != 0) {
+        return PyErr_Format(PyExc_TypeError, "GenericAlias does not support keyword arguments");
+    }
+    if (PyTuple_Size(args) != 2) {
```
```suggestion
    if (PyTuple_GET_SIZE(args) != 2) {
```

gvanrossum committed:
Use __class_getitem__ = classmethod(GenericAlias)

gvanrossum committed:
Rewrite ga_repr_item() to honor exceptions

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>

gvanrossum committed:
Use PyTuple_GET_SIZE

gvanrossum committed:
Update Objects/genericaliasobject.c

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>

gvanrossum committed:
Apply suggestions from code review

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>

gvanrossum committed:
Use PyErr_SetString correctly

gvanrossum committed:
Fix test failure (setattr may raise TypeError)

gvanrossum committed:
Propagate errors from is_typevar()

serhiy-storchaka commented on .github/workflows/build.yml:
```diff
@@ -13,6 +13,7 @@ on:
     - '**/*.rst'
   pull_request:
     branches:
+    - pep585
```
Is it needed?

gvanrossum replied:
Only as long as we're using this branch for the PR. I'll remove it after merging master once last time.

serhiy-storchaka commented on Objects/genericaliasobject.c:
```diff
@@ -36,54 +36,86 @@ ga_traverse(PyObject *self, visitproc visit, void *arg)
+    {
+        // builtins don't need a module name
+        Py_INCREF(qualname);
+        r = qualname;
```
```suggestion
        r = PyObject_Str(qualname);
```
`_PyUnicodeWriter_WriteStr()` will crash for a non-string.

serhiy-storchaka reviewed (approved):
One last fix (for `__qualname__`) and LGTM. :+1:

gvanrossum committed:
Use PyObject_Str(qualname)

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>

gvanrossum committed:
Merge remote-tracking branch 'upstream/master' into pep585

gvanrossum committed:
Delete mention of pep585 branch in GitHub workflow

gvanrossum committed:
Move Py_GenericAlias* to genericaliasobject.h

gvanrossum committed:
Merge remote-tracking branch 'upstream/master' into pep585

gvanrossum committed:
Add Py_GenericAlias[Type] to PC/python3.ref
